src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -531,6 +538,13 @@ private void ProcessEvents()
                             previousEventCookie = 0;
                         }
 
+                        // If the event signaled that there's a new subdirectory and if we're monitoring subdirectories,
+                        // add a watch for it.
+                        if (addWatch && isDir && _includeSubdirectories)
+                        {
+                            AddDirectoryWatch(associatedDirectoryEntry, nextEvent.name);
+                        }
+
```

Comment:
The bug was that the Creation event would be fired to clients before adding the watch for the folder; this meant that your previous comment about waiting for the creation event before continuing (to get rid of the Thread.Sleep) would be incorrect since the race condition would still occur since the event would be fired and the UT would proceed before the FSW had a chance to add the new watch